File: backend/app/routers/user_management.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db

from app.models.user import User

from app.schemas.user import UserCreate, ROLES

from app.utils.permissions import (
    require_roles,
    ADMINISTRATOR
)

from app.utils.security import hash_password


router = APIRouter(
    prefix="/user-management",
    tags=["User Management"]
)


# ==========================================
# GET ALL USERS
# ==========================================

@router.get("/")
def get_all_users(
    db: Session = Depends(get_db),

    current_user=Depends(
        require_roles(
            ADMINISTRATOR
        )
    )
):

    users = (
        db.query(User)
        .order_by(
            User.id.asc()
        )
        .all()
    )


    return [

        {
            "id": user.id,
            "full_name": user.full_name,
            "email": user.email,
            "role": user.role
        }

        for user in users

    ]


# ==========================================
# GET SINGLE USER
# ==========================================

@router.get("/{user_id}")
def get_user(
    user_id: int,

    db: Session = Depends(get_db),

    current_user=Depends(
        require_roles(
            ADMINISTRATOR
        )
    )
):

    user = (
        db.query(User)
        .filter(
            User.id == user_id
        )
        .first()
    )


    if not user:

        raise HTTPException(
            status_code=404,
            detail="User not found"
        )


    return {

        "id": user.id,

        "full_name":
            user.full_name,

        "email":
            user.email,

        "role":
            user.role

    }


# ==========================================
# CREATE USER
# ==========================================

@router.post("/")
def create_user(
    data: UserCreate,

    db: Session = Depends(get_db),

    current_user=Depends(
        require_roles(
            ADMINISTRATOR
        )
    )
):

    # ======================================
    # VALIDATE ROLE
    # ======================================

    if data.role not in ROLES:

        raise HTTPException(
            status_code=400,
            detail="Invalid user role"
        )


    # ======================================
    # CHECK EMAIL
    # ======================================

    existing_user = (
        db.query(User)
        .filter(
            User.email == data.email
        )
        .first()
    )


    if existing_user:

        raise HTTPException(
            status_code=400,
            detail="Email already registered"
        )


    # ======================================
    # CREATE USER
    # ======================================

    user = User(

        full_name=data.full_name,

        email=data.email,

        password=
            hash_password(
                data.password
            ),

        role=data.role

    )


    db.add(user)

    db.commit()

    db.refresh(user)


    # ======================================
    # RESPONSE
    # ======================================

    return {

        "message":
            "User created successfully",

        "user": {

            "id":
                user.id,

            "full_name":
                user.full_name,

            "email":
                user.email,

            "role":
                user.role

        }

    }


# ==========================================
# UPDATE USER ROLE
# ==========================================

@router.put("/{user_id}/role")
def update_user_role(
    user_id: int,

    role_data: dict,

    db: Session = Depends(get_db),

    current_user=Depends(
        require_roles(
            ADMINISTRATOR
        )
    )
):

    # ======================================
    # GET USER
    # ======================================

    user = (
        db.query(User)
        .filter(
            User.id == user_id
        )
        .first()
    )


    if not user:

        raise HTTPException(
            status_code=404,
            detail="User not found"
        )


    # ======================================
    # GET ROLE
    # ======================================

    new_role = role_data.get(
        "role"
    )


    if not new_role:

        raise HTTPException(
            status_code=400,
            detail="Role is required"
        )


    # ======================================
    # VALIDATE ROLE
    # ======================================

    if new_role not in ROLES:

        raise HTTPException(
            status_code=400,
            detail="Invalid user role"
        )


    # ======================================
    # UPDATE ROLE
    # ======================================

    user.role = new_role


    db.commit()

    db.refresh(user)


    return {

        "message":
            "User role updated successfully",

        "user": {

            "id":
                user.id,

            "full_name":
                user.full_name,

            "email":
                user.email,

            "role":
                user.role

        }

    }


# ==========================================
# UPDATE USER
# ==========================================

@router.put("/{user_id}")
def update_user(
    user_id: int,

    data: dict,

    db: Session = Depends(get_db),

    current_user=Depends(
        require_roles(
            ADMINISTRATOR
        )
    )
):

    # ======================================
    # GET USER
    # ======================================

    user = (
        db.query(User)
        .filter(
            User.id == user_id
        )
        .first()
    )


    if not user:

        raise HTTPException(
            status_code=404,
            detail="User not found"
        )


    # ======================================
    # FULL NAME
    # ======================================

    if "full_name" in data:

        full_name = str(
            data["full_name"]
        ).strip()


        if len(full_name) < 2:

            raise HTTPException(
                status_code=400,
                detail="Full name must contain at least 2 characters"
            )


        user.full_name = full_name


    # ======================================
    # EMAIL
    # ======================================

    if "email" in data:

        email = str(
            data["email"]
        ).strip()


        existing_user = (
            db.query(User)
            .filter(
                User.email == email,
                User.id != user_id
            )
            .first()
        )


        if existing_user:

            raise HTTPException(
                status_code=400,
                detail="Email already registered"
            )


        user.email = email


    # ======================================
    # ROLE
    # ======================================

    if "role" in data:

        new_role = data["role"]


        if new_role not in ROLES:

            raise HTTPException(
                status_code=400,
                detail="Invalid user role"
            )


        user.role = new_role


    # ======================================
    # PASSWORD
    # ======================================

    if data.get("password"):

        password = str(
            data["password"]
        )


        if len(password) < 8:

            raise HTTPException(
                status_code=400,
                detail="Password must contain at least 8 characters"
            )


        user.password = hash_password(
            password
        )


    # ======================================
    # SAVE
    # ======================================

    db.commit()

    db.refresh(user)


    return {

        "message":
            "User updated successfully",

        "user": {

            "id":
                user.id,

            "full_name":
                user.full_name,

            "email":
                user.email,

            "role":
                user.role

        }

    }


# ==========================================
# DELETE USER
# ==========================================

@router.delete("/{user_id}")
def delete_user(
    user_id: int,

    db: Session = Depends(get_db),

    current_user=Depends(
        require_roles(
            ADMINISTRATOR
        )
    )
):

    # ======================================
    # PREVENT SELF DELETE
    # ======================================

    if current_user.id == user_id:

        raise HTTPException(
            status_code=400,
            detail="You cannot delete your own account"
        )


    # ======================================
    # GET USER
    # ======================================

    user = (
        db.query(User)
        .filter(
            User.id == user_id
        )
        .first()
    )


    if not user:

        raise HTTPException(
            status_code=404,
            detail="User not found"
        )


    # ======================================
    # DELETE
    # ======================================

    db.delete(user)

    db.commit()


    return {

        "message":
            "User deleted successfully"

    }

File: frontend/src/app/pages/user-management/user-management.ts
import {
  Component,
  OnInit,
  signal
} from '@angular/core';

import {
  CommonModule
} from '@angular/common';

import {
  FormsModule
} from '@angular/forms';

import {
  UserManagement as UserManagementService
} from '../../services/user-management';

import {
  Auth
} from '../../services/auth';


@Component({
  selector: 'app-user-management',

  standalone: true,

  imports: [
    CommonModule,
    FormsModule
  ],

  templateUrl:
    './user-management.html',

  styleUrl:
    './user-management.css'
})
export class UserManagement implements OnInit {


  // ==========================================
  // USERS
  // ==========================================

  users =
    signal<any[]>([]);


  // ==========================================
  // LOADING
  // ==========================================

  loading =
    signal(false);


  saving =
    signal(false);


  // ==========================================
  // FORM
  // ==========================================

  showForm =
    signal(false);


  editingUserId:
    number | null = null;


  fullName =
    '';


  email =
    '';


  password =
    '';


  role =
    'Vendor';


  // ==========================================
  // ROLES
  // ==========================================

  roles = [

    'Administrator',

    'Procurement Manager',

    'Supply Chain Manager',

    'Vendor',

    'Finance Officer',

    'Auditor'

  ];


  // ==========================================
  // CURRENT USER
  // ==========================================

  currentUser:
    any = null;


  // ==========================================
  // CONSTRUCTOR
  // ==========================================

  constructor(

    private userService:
      UserManagementService,

    private auth:
      Auth

  ) {}


  // ==========================================
  // INITIALIZE
  // ==========================================

  ngOnInit(): void {

    this.currentUser =
      this.auth.getCurrentUser();

    this.loadUsers();

  }


  // ==========================================
  // LOAD USERS
  // ==========================================

  loadUsers(): void {

    this.loading.set(true);


    this.userService
      .getUsers()
      .subscribe({

        next: (response: any[]) => {

          this.users.set(
            response
          );

          this.loading.set(false);

        },

        error: (error: any) => {

          console.error(
            'Failed to load users:',
            error
          );

          alert(
            error?.error?.detail ||
            'Failed to load users.'
          );

          this.users.set([]);

          this.loading.set(false);

        }

      });

  }


  // ==========================================
  // OPEN CREATE FORM
  // ==========================================

  openCreateForm(): void {

    this.resetForm();

    this.showForm.set(true);

  }


  // ==========================================
  // OPEN EDIT FORM
  // ==========================================

  openEditForm(
    user: any
  ): void {

    this.editingUserId =
      user.id;

    this.fullName =
      user.full_name;

    this.email =
      user.email;

    this.password =
      '';

    this.role =
      user.role;

    this.showForm.set(true);

  }


  // ==========================================
  // CLOSE FORM
  // ==========================================

  closeForm(): void {

    this.showForm.set(false);

    this.resetForm();

  }


  // ==========================================
  // RESET FORM
  // ==========================================

  resetForm(): void {

    this.editingUserId =
      null;

    this.fullName =
      '';

    this.email =
      '';

    this.password =
      '';

    this.role =
      'Vendor';

  }


  // ==========================================
  // SAVE USER
  // ==========================================

  saveUser(): void {

    // ========================================
    // VALIDATION
    // ========================================

    if (
      !this.fullName.trim()
    ) {

      alert(
        'Please enter full name.'
      );

      return;

    }


    if (
      !this.email.trim()
    ) {

      alert(
        'Please enter email.'
      );

      return;

    }


    if (
      !this.editingUserId &&
      !this.password
    ) {

      alert(
        'Please enter password.'
      );

      return;

    }


    if (
      this.password &&
      this.password.length < 8
    ) {

      alert(
        'Password must contain at least 8 characters.'
      );

      return;

    }


    // ========================================
    // CREATE
    // ========================================

    if (
      this.editingUserId === null
    ) {

      const data = {

        full_name:
          this.fullName.trim(),

        email:
          this.email.trim(),

        password:
          this.password,

        role:
          this.role

      };


      this.saving.set(true);


      this.userService
        .createUser(data)
        .subscribe({

          next: () => {

            alert(
              'User created successfully.'
            );

            this.saving.set(false);

            this.closeForm();

            this.loadUsers();

          },

          error: (error: any) => {

            console.error(
              'Create user error:',
              error
            );

            this.saving.set(false);

            alert(
              error?.error?.detail ||
              'Failed to create user.'
            );

          }

        });


      return;

    }


    // ========================================
    // UPDATE
    // ========================================

    const data: any = {

      full_name:
        this.fullName.trim(),

      email:
        this.email.trim(),

      role:
        this.role

    };


    if (
      this.password.trim()
    ) {

      data.password =
        this.password;

    }


    this.saving.set(true);


    this.userService
      .updateUser(
        this.editingUserId,
        data
      )
      .subscribe({

        next: () => {

          alert(
            'User updated successfully.'
          );

          this.saving.set(false);

          this.closeForm();

          this.loadUsers();

        },

        error: (error: any) => {

          console.error(
            'Update user error:',
            error
          );

          this.saving.set(false);

          alert(
            error?.error?.detail ||
            'Failed to update user.'
          );

        }

      });

  }


  // ==========================================
  // DELETE USER
  // ==========================================

  deleteUser(
    user: any
  ): void {

    if (
      this.currentUser?.id ===
      user.id
    ) {

      alert(
        'You cannot delete your own account.'
      );

      return;

    }


    const confirmed =
      confirm(
        `Are you sure you want to delete ${user.full_name}?`
      );


    if (!confirmed) {

      return;

    }


    this.userService
      .deleteUser(
        user.id
      )
      .subscribe({

        next: () => {

          alert(
            'User deleted successfully.'
          );

          this.loadUsers();

        },

        error: (error: any) => {

          console.error(
            'Delete user error:',
            error
          );

          alert(
            error?.error?.detail ||
            'Failed to delete user.'
          );

        }

      });

  }


  // ==========================================
  // GET ROLE CLASS
  // ==========================================

  getRoleClass(
    role: string
  ): string {

    switch (role) {

      case 'Administrator':
        return 'role-admin';

      case 'Procurement Manager':
        return 'role-procurement';

      case 'Supply Chain Manager':
        return 'role-supply';

      case 'Vendor':
        return 'role-vendor';

      case 'Finance Officer':
        return 'role-finance';

      case 'Auditor':
        return 'role-auditor';

      default:
        return '';

    }

  }

}